### PR TITLE
Add request body reader

### PR DESCRIPTION
**Pull request by cesarbs**
Will be used in a subsequent change to add a request body timeout/minimum data rate.

### CODE REVIEW

**Comment by cesarbs** [comment]
Perf

**dev**

```
[11:06:19.457] Starting scenario Plaintext on benchmark server...
[11:07:27.435] Warmup
[11:07:27.444] Starting scenario Plaintext on benchmark client...
[11:07:43.550] Scenario Plaintext completed on benchmark client
[11:07:43.551] RPS: 1173022.98
[11:07:43.552] Stopping scenario Plaintext on benchmark client...
[11:07:43.554] Benchmark
[11:07:43.557] Starting scenario Plaintext on benchmark client...
[11:07:59.595] Scenario Plaintext completed on benchmark client
[11:07:59.595] RPS: 1225692.46
[11:07:59.596] Stopping scenario Plaintext on benchmark client...
[11:07:59.601] Stopping scenario Plaintext on benchmark server...
```

**PR**

```
[11:02:15.858] Starting scenario Plaintext on benchmark server...
[11:03:03.382] Warmup
[11:03:03.390] Starting scenario Plaintext on benchmark client...
[11:03:19.518] Scenario Plaintext completed on benchmark client
[11:03:19.518] RPS: 1163631.37
[11:03:19.519] Stopping scenario Plaintext on benchmark client...
[11:03:19.521] Benchmark
[11:03:19.521] Starting scenario Plaintext on benchmark client...
[11:03:35.557] Scenario Plaintext completed on benchmark client
[11:03:35.557] RPS: 1200649.98
[11:03:35.559] Stopping scenario Plaintext on benchmark client...
[11:03:35.561] Stopping scenario Plaintext on benchmark server...
```

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                    }
+                }
+                finally
+                {
+                    _pipe.Reader.Advance(readableBuffer.End);
+                }
+            }
+        }
+
+        public Task ConsumeAsync(CancellationToken cancellationToken = default(CancellationToken))
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs]: Did you really need this split?

Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs]: Seemed to help a little when I was fixing perf. And we already had something like that in `MessageBody`, so why not?

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/IRequestBodyReader.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
+{
+    public interface IRequestBodyReader
+    {
+        Task StartAsync(MessageBody messageBody, CancellationToken cancellationToken = default(CancellationToken));
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/IRequestBodyReader.cs]: Might be better to make cancellationToken a required parameter. It will save us from accidentally not passing the token to methods as we work on this code.

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        public static readonly IRequestBodyReader Instance = new EmptyRequestBodyReader();
+
+        public Task ConsumeAsync(CancellationToken cancellationToken = default(CancellationToken))
+            => TaskCache.CompletedTask;
+
+        public Task CopyToAsync(Stream destination, CancellationToken cancellationToken = default(CancellationToken))
+            => TaskCache.CompletedTask;
+
+        public Task<int> ReadAsync(ArraySegment<byte> buffer, CancellationToken cancellationToken = default(CancellationToken))
+            => Task.FromResult(0);
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs]: Nit: use `TaskCache<int>.DefaultCompletedTask` or cache this task on a static.

File: test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/FrameResponseHeadersTests.cs [hunk 4]
```diff
@@ -59,14 +64,14 @@ public void InitialDictionaryIsEmpty()
...
-        [InlineData("Unknown�-Header", "Data")]
-        [InlineData("Ser�ver", "Data")]
+        [InlineData("Ser™ver", "Data")]
+        [InlineData("Server", "Da™ta")]
+        [InlineData("Unknown™-Header", "Data")]
+        [InlineData("Ser™ver", "Data")]
+        [InlineData("šerver", "Data")]
+        [InlineData("Server", "Dašta")]
+        [InlineData("Unknownš-Header", "Data")]
+        [InlineData("Seršver", "Data")]
```
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/FrameResponseHeadersTests.cs]: What changed here? On GitHub the InlineData looks the same before and after.

Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/FrameResponseHeadersTests.cs]: I got a bunch of question marks at first. Then I copied & pasted from the original and got this. Need to get rid of the diff.

Comment (natemcmaster) [inline test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/FrameResponseHeadersTests.cs]: File -> Save As... > Save 🔽  > Save with Encoding...

Actually, I suspect the new encoding is right, but good to double check that its utf8.

image

Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/FrameResponseHeadersTests.cs]: Was extended ASCII, is now UTF-8.

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+                {
+                    return;
+                }
+
+                awaitable = _pipe.Reader.ReadAsync();
+            }
+        }
+    }
+
+    public class EmptyRequestBodyReader : IRequestBodyReader
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs]: Separate file

**Comment by cesarbs** [comment]
Please look at the update to the first comment wrt reset.

File: test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/FrameResponseHeadersTests.cs (hunk 4 above)
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/FrameResponseHeadersTests.cs]: Not intentional, but I assume it all ends up the same since the compiler would have to convert the string literals to UTF-16 either way.

Comment (benaadams) [inline test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/FrameResponseHeadersTests.cs]: If you save wrong it it can easily change; in coreclr they force use of surrogate pairs \uxxx to avoid the risk (VS seems to like using non extended encodings) - which made me sad as I had originally used a bunch of emojis 😞 

However it should stand out on a PR; so is probably ok?

**Comment by davidfowl** [comment]
@cesarbs Can you run the Copy benchmark?

**Comment by davidfowl** [comment]
https://github.com/aspnet/benchmarks/blob/52570f73bf0ab038e4de3c123c367c072ecfd2c0/src/Benchmarks/Configuration/Scenarios.cs#L58

**Comment by cesarbs** [comment]
@davidfowl Copy benchmark is surpringly unaffected:

**dev**

```
[06:46:31.545] Starting scenario Copy on benchmark server...
[06:47:21.303] Warmup
[06:47:21.310] Starting scenario Copy on benchmark client...
[06:47:36.536] Scenario Copy completed on benchmark client
[06:47:36.536] RPS: 128780.48
[06:47:36.540] Stopping scenario Copy on benchmark client...
[06:47:36.542] Benchmark
[06:47:36.543] Starting scenario Copy on benchmark client...
[06:47:52.792] Scenario Copy completed on benchmark client
[06:47:52.793] RPS: 136166.23
[06:47:52.794] Stopping scenario Copy on benchmark client...
[06:47:52.796] Stopping scenario Copy on benchmark server...
```

**PR**

```
[06:48:11.754] Starting scenario Copy on benchmark server...
[06:48:52.673] Warmup
[06:48:52.680] Starting scenario Copy on benchmark client...
[06:49:07.970] Scenario Copy completed on benchmark client
[06:49:07.971] RPS: 134193.58
[06:49:07.972] Stopping scenario Copy on benchmark client...
[06:49:07.974] Benchmark
[06:49:07.974] Starting scenario Copy on benchmark client...
[06:49:24.267] Scenario Copy completed on benchmark client
[06:49:24.267] RPS: 136440.26
[06:49:24.269] Stopping scenario Copy on benchmark client...
[06:49:24.272] Stopping scenario Copy on benchmark server...
```

**Comment by davidfowl** [comment]
👏

**Comment by cesarbs** [comment]
Previous numbers were wrong due to my wrong assumption that the "copy" benchmark job would send data to the server. I've changed that: https://github.com/aspnet/benchmarks/pull/261

Measurements from my VMs:

**dev**

```
[07:12:01.528] Starting scenario CopyToAsync on benchmark server...
[07:12:39.604] Warmup
[07:12:39.611] Starting scenario CopyToAsync on benchmark client...
[07:12:55.500] Scenario CopyToAsync completed on benchmark client
[07:12:55.500] RPS: 2777.74
[07:12:55.501] Stopping scenario CopyToAsync on benchmark client...
[07:12:55.503] Benchmark
[07:12:55.503] Starting scenario CopyToAsync on benchmark client...
[07:13:10.996] Scenario CopyToAsync completed on benchmark client
[07:13:10.996] RPS: 2824.38
[07:13:10.997] Stopping scenario CopyToAsync on benchmark client...
[07:13:10.999] Stopping scenario CopyToAsync on benchmark server...
```

**PR**

```
[07:13:24.491] Starting scenario CopyToAsync on benchmark server...
[07:13:42.188] Warmup
[07:13:42.195] Starting scenario CopyToAsync on benchmark client...
[07:13:57.441] Scenario CopyToAsync completed on benchmark client
[07:13:57.441] RPS: 2779.84
[07:13:57.442] Stopping scenario CopyToAsync on benchmark client...
[07:13:57.444] Benchmark
[07:13:57.444] Starting scenario CopyToAsync on benchmark client...
[07:14:12.682] Scenario CopyToAsync completed on benchmark client
[07:14:12.682] RPS: 2805.55
[07:14:12.683] Stopping scenario CopyToAsync on benchmark client...
[07:14:12.685] Stopping scenario CopyToAsync on benchmark server...
```

Numbers still look similar, but are much lower compared to plaintext given 100K of data is being sent on every request. To confirm that everything was working alright, I enabled logging on the server and watched as requests came in from wrk, and confirmed they had a body and were being read.

I believe the reason the numbers are similar is that the request reader runs concurrent to the application, and starts before it as well. That's giving it enough time to read some body data before the app requests it, and it keeps happening while the app keeps requesting it (i.e. it's always reading from the network ahead of the app). This would make the cost of an app read not much different from a read from `MessageBody` before this change, assuming data is available.

I'm removing the WIP tag from the title.

**Comment by cesarbs** [comment]
Numbers from our perf lab:

**dev**

```
[12:27:32.899] Starting scenario CopyToAsync on benchmark server...
[12:28:25.770] Warmup
[12:28:25.775] Starting scenario CopyToAsync on benchmark client...
[12:28:41.887] Scenario CopyToAsync completed on benchmark client
[12:28:41.888] RPS: 10469.92
[12:28:41.889] Stopping scenario CopyToAsync on benchmark client...
[12:28:41.891] Benchmark
[12:28:41.891] Starting scenario CopyToAsync on benchmark client...
[12:28:57.941] Scenario CopyToAsync completed on benchmark client
[12:28:57.941] RPS: 10566.76
[12:28:57.942] Stopping scenario CopyToAsync on benchmark client...
[12:28:57.944] Stopping scenario CopyToAsync on benchmark server...
```

**PR**

```
[12:25:52.141] Starting scenario CopyToAsync on benchmark server...
[12:26:52.655] Warmup
[12:26:52.661] Starting scenario CopyToAsync on benchmark client...
[12:27:08.777] Scenario CopyToAsync completed on benchmark client
[12:27:08.777] RPS: 10340.61
[12:27:08.778] Stopping scenario CopyToAsync on benchmark client...
[12:27:08.780] Benchmark
[12:27:08.780] Starting scenario CopyToAsync on benchmark client...
[12:27:24.822] Scenario CopyToAsync completed on benchmark client
[12:27:24.822] RPS: 10499.63
[12:27:24.823] Stopping scenario CopyToAsync on benchmark client...
[12:27:24.825] Stopping scenario CopyToAsync on benchmark server...
```

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                    {
+                        bytesRead = await messageBody.ReadAsync(writableBuffer.Buffer.GetArray(), cancellationToken);
+                        writableBuffer.Advance(bytesRead);
+                    }
+                    finally
+                    {
+                        writableBuffer.Commit();
+                    }
+
+                    if (bytesRead == 0)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs]: I prefer this pattern:

https://github.com/aspnet/KestrelHttpServer/blob/37f15bdd85cf9e32904232554202dc4a6052ef42/src/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/SocketConnection.cs#L87-L102

**Comment by cesarbs** [comment]
📢 Any more feedback?

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs [hunk 7]
```diff
@@ -30,6 +30,8 @@ protected MessageBody(Frame context)
 
         public bool RequestUpgrade { get; protected set; }
 
+        public virtual bool IsEmpty => false;
+
         public Task<int> ReadAsync(ArraySegment<byte> buffer, CancellationToken cancellationToken = default(CancellationToken))
         {
             var task = PeekAsync(cancellationToken);
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs]: Not necessary as part of this PR, but I agree with @pakrym that PeekAsync can be changed back to ReadAsync now that CopyToAsync doesn't use it.

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs [hunk 8]
```diff
@@ -68,145 +70,7 @@ private int CopyReadSegment(ArraySegment<byte> readSegment, ArraySegment<byte> b
...
-                    return;
-                }
-
-                ConsumedBytes(count);
-                currentTask = PeekAsync(cancellationToken);
-            }
-        }
-
-        private void TryProduceContinue()
+        public void TryProduceContinue()
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs]: Why did you change this from private to public? This is only being called by MessageBody.ReadAsync(), right?

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+                    }
+                }
+                finally
+                {
+                    _pipe.Reader.Advance(readableBuffer.End);
+                }
+            }
+        }
+
+        public virtual async Task CopyToAsync(Stream destination, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/RequestBodyReader.cs]: No more CopyToAsync(Read|Destination)Awaited? 😆

File: test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/RequestBodyReaderTests.cs [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+                var pipe = input.Pipe;
+                var bytes = Encoding.ASCII.GetBytes(data[0]);
+                var buffer = pipe.Writer.Alloc(2048);
+                Assert.True(buffer.Buffer.TryGetArray(out var block));
+                Buffer.BlockCopy(bytes, 0, block.Array, block.Offset, bytes.Length);
+                buffer.Advance(bytes.Length);
+                await buffer.FlushAsync();
+
+                // Verify the block passed to WriteAsync is the same one incoming data was written into.
+                Assert.Same(block.Array, await writeTcs.Task);
```
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/RequestBodyReaderTests.cs]: I think this test is passing only because you're using the same PipeFactory and therefore the same MemoryPool for  input.Pipe and RequestBodyReader._pipe. Since both Pipes are backed by the same pool, the data copied from input.Pipe to the writable buffer of RequestBodyReader._pipe ends up in the same array instance.

After this change, CopyToAsyncDoes~~Not~~CopyBlocks, so maybe this test should just be removed.

File: src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs (hunk 7 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs]: I'm already working on that change, but I'll send a separate PR for it.

Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs]: After discussing in person, I like returning a ReadableBuffer and appending that to the RequestBodyReader pipe even more. That way we can even keep the CopyToAsyncDoesNotCopyBlocks test 😄

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs]: 👍

**Comment by cesarbs** [comment]
Closing in favor of #1840.
